// @transia/ripple-binary-codec/dist/types/dataType.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.DataType = void 0;
const serialized_type_1 = require("./serialized-type");
const utils_1 = require("../utils");
/**
 * STDataType: Encodes XRPL's "DataType" type.
 *
 * This type wraps an inner SerializedTypeID to indicate what type of data
 * a field contains. It's encoded as a 2-byte unsigned integer representing
 * the inner type.
 *
 * Usage:
 *   DataType.from({ type: "AMOUNT" })
 *   DataType.from("UINT64")
 *   DataType.fromParser(parser)
 */
class DataType extends serialized_type_1.SerializedType {
    /**
     * Construct a DataType from bytes
     * @param bytes - 2-byte Uint8Array containing the inner type ID
     * @param innerType - Optional explicit inner type (used when constructing from value)
     * @throws Error if bytes is not a 2-byte Uint8Array
     */
    constructor(bytes, innerType) {
        const used = bytes !== null && bytes !== void 0 ? bytes : DataType.defaultBytes;
        if (!(used instanceof Uint8Array) || used.length !== 2) {
            throw new Error(`DataType must be constructed from a 2-byte Uint8Array, got ${used === null || used === void 0 ? void 0 : used.length} bytes`);
        }
        super(used);
        // If innerType is explicitly provided, use it; otherwise read from bytes
        if (innerType !== undefined) {
            this.innerType = innerType;
        }
        else {
            this.innerType = (0, utils_1.readUInt16BE)(used, 0);
        }
    }
    /**
     * Construct from various input types
     *
     * @param value - Can be:
     *   - DataType instance (returns as-is)
     *   - DataTypeJSON object with 'type' field
     *   - String type name (e.g., "AMOUNT", "UINT64")
     *   - SerializedTypeID enum value
     * @returns DataType instance
     * @throws Error if value type is not supported or type string is unknown
     */
    static from(value) {
        if (value instanceof DataType) {
            return value;
        }
        if (typeof value === 'object' && value !== null && 'type' in value) {
            const json = value;
            return DataType.fromTypeString(json.type);
        }
        if (typeof value === 'string') {
            return DataType.fromTypeString(value);
        }
        if (typeof value === 'number') {
            return DataType.fromTypeId(value);
        }
        throw new Error('DataType.from: value must be DataType, DataTypeJSON, string, or SerializedTypeID');
    }
    /**
     * Construct from a type string
     *
     * @param typeStr - Type string like "AMOUNT", "UINT64", etc.
     * @returns DataType instance
     * @throws Error if type string is not recognized
     */
    static fromTypeString(typeStr) {
        const typeId = serialized_type_1.TYPE_STRING_TO_ID[typeStr];
        if (typeId === undefined) {
            throw new Error(`DataType: unsupported type string: ${typeStr}`);
        }
        return DataType.fromTypeId(typeId);
    }
    /**
     * Construct from a SerializedTypeID
     *
     * @param typeId - The SerializedTypeID enum value
     * @returns DataType instance
     */
    static fromTypeId(typeId) {
        const bytes = new Uint8Array(2);
        (0, utils_1.writeUInt16BE)(bytes, typeId, 0);
        return new DataType(bytes, typeId);
    }
    /**
     * Read a DataType from a BinaryParser stream (2 bytes)
     *
     * @param parser - BinaryParser positioned at the start of a DataType
     * @returns DataType instance
     */
    static fromParser(parser) {
        const bytes = parser.read(2);
        return new DataType(bytes);
    }
    /**
     * Get the inner SerializedTypeID
     *
     * @returns The inner type ID
     */
    getInnerType() {
        return this.innerType;
    }
    /**
     * Set the inner SerializedTypeID
     *
     * @param typeId - The new inner type ID
     */
    setInnerType(typeId) {
        this.innerType = typeId;
        (0, utils_1.writeUInt16BE)(this.bytes, typeId, 0);
    }
    /**
     * Get the string representation of the inner type
     *
     * @returns String name of the type, or numeric string if unknown
     */
    getInnerTypeString() {
        return serialized_type_1.TYPE_ID_TO_STRING[this.innerType] || this.innerType.toString();
    }
    /**
     * Convert to JSON representation
     *
     * @returns JSON object with 'type' field
     */
    toJSON() {
        return {
            type: this.getInnerTypeString(),
        };
    }
    getSType() {
        return serialized_type_1.SerializedTypeID.STI_DATATYPE;
    }
}
exports.DataType = DataType;
/**
 * Default bytes for DataType (STI_NOTPRESENT)
 */
DataType.defaultBytes = new Uint8Array([0x00, 0x01]);
